### ISSUE TITLE
Lacking a .QuerySingle() makes us break the async conventions

### ISSUE THREAD

**Issue by DumboJetEngine** [StackExchange/dapper-dot-net#260]
I am reporting a dilemma that I am facing right now. Perhaps I am missing something or perhaps not.

I was creating a class of my own that used Dapper internally to make DB queries.
I wanted to expose a set of `*Async()` methods and I came across this case of non-async code:

```
        public TUser GetUserById(int memberId)
        {
            return db.Connection.Query<TUser>("Select * from Member where Id=@MemberId", new { MemberId = memberId })
                .FirstOrDefault();
        }
```
Now, if I am to make an async method out of this and keep the usual conventions (e.g. that async methods return `Task<T>` or `void`), how can I implement a method with this signature? :

```
public async Task<TUser> GetUserByIdAsync(int memberId);
```
As far as I saw, there is no method returning a single row from a query.
Correct me if I am wrong, but I think that `ExecuteScalarAsync()` is meant to fetch scalar values, not rows.
Also, using `QueryAsync()`, returns a `Task<IEnumerable<TUser>>`, which is not exactly what I want my method to return.
I could use await in my method and the use `.FirstOrDefault()` on the result, but then my method would return `TUser` and not `Task<TUser>` that is awaitable.
Also, I am not sure if I could somehow re-wrap my single result into a `Task<TUser>` and, if I could, I am not sure this is the correct way to do things.

Any ideas or hints would be welcome.

**Comment by DumboJetEngine** [comment]
Do you think that this is an appropriate implementation? :

```
        public async Task<TUser> GetUserByIdAsync(int memberId)
        {
            var res = await db.Connection.QueryAsync<TUser>("Select * from Member where Id=@MemberId", new {MemberId = memberId});
            return res.FirstOrDefault();
        }
```

This is my first SERIOUS work with async, so I am a bit skeptical about things...
Perhaps this should be posted on StackOverflow instead...

**Comment by mgravell** [comment]
Yes, basically; that should be fine - or even just:

    return (await db.Connection.blahblahblah).FirstOrDefault();

On 25 March 2015 at 14:33, DumboJetEngine <notifications@github.com> wrote:

> Closed #260 <https://github.com/StackExchange/dapper-dot-net/issues/260>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/dapper-dot-net/issues/260#event-264773418>
> .
>



-- 
Regards,

Marc

**Comment by DumboJetEngine** [comment]
Thanks! :thumbsup:
